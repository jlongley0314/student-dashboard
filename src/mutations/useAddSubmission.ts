import { useMutation } from "react-query";
import { Submission } from "../types";

export function useAddSubmission(id: string) {
  return useMutation(async (submission: Omit<Submission, "id">) => {
    const response = await fetch(
      `https://613bb521110e000017a4570e.mockapi.io/students/${id}/submissions`,
      {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(submission),
      }
    );

    if (!response.ok) {
      const error = await response.json();
      console.log("error", error);
      throw error.error;
    }

    return response;
  });
}
