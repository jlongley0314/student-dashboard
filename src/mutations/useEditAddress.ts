import { useMutation } from "react-query";

export function useEditAddress(id: string) {
  return useMutation(
    async (formData: FormData) => {
      const response = await fetch(
        `https://613bb521110e000017a4570e.mockapi.io/students/${id}/address`,
        {
          method: "POST",
          body: formData,
        }
      );

      if (!response.ok) {
        const error = await response.json();
        console.log("error", error);
        throw error.error;
      }

      return response;
    },
    { onSuccess: () => {} }
  );
}
