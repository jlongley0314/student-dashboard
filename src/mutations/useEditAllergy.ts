import { useMutation } from "react-query";
import { Allergy } from "../types";

export function useEditAllergy(id: string) {
  return useMutation(async (allergy: Allergy) => {
    const response = await fetch(
      `https://613bb521110e000017a4570e.mockapi.io/students/${id}/allergies/${allergy.id}`,
      {
        method: "PUT",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(allergy),
      }
    );

    if (!response.ok) {
      const error = await response.json();
      console.log("error", error);
      throw error.error;
    }

    return response;
  });
}
