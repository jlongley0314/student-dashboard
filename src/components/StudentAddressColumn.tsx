import React, { useState } from "react";
import { Address } from "../types";
import {
  Card,
  Typography,
  Stack,
  Button,
  TextField,
  Box,
  LinearProgress,
} from "@mui/material";
import { useEditAddress } from "../mutations/useEditAddress";

type StudentAddressColumnProps = {
  address: Address;
  id: string;
};

export function StudentAddressColumn(props: StudentAddressColumnProps) {
  const { address, id } = props;
  const [isEditing, setIsEditing] = useState(false);
  const [formData, setFormData] = useState(address);
  const editAddressMutation = useEditAddress(id);

  function handleClick() {
    setIsEditing(!isEditing);
    if (isEditing) {
      editAddressMutation.mutate(formData);
    }
  }

  return (
    <Card variant="outlined" style={{ padding: 10 }}>
      {editAddressMutation.isLoading && <LinearProgress />}
      <Typography variant="h5" component="div" gutterBottom>
        Address:
      </Typography>
      {isEditing ? (
        <Box
          component="form"
          sx={{
            "& .MuiTextField-root": { m: 1, width: "50ch" },
          }}
          noValidate
          autoComplete="off"
        >
          <div>
            <TextField
              required
              error={!formData.line1}
              label="Address Line 1"
              defaultValue={address.line1}
              onChange={(event) => {
                setFormData({ ...formData, line1: event.target.value });
              }}
            />
            <TextField
              label="Address Line 2"
              defaultValue={address.line2}
              onChange={(event) => {
                setFormData({ ...formData, line2: event.target.value });
              }}
            />
            <TextField
              required
              error={!formData.city}
              label="City"
              defaultValue={address.city}
              onChange={(event) => {
                setFormData({ ...formData, city: event.target.value });
              }}
            />
            <TextField
              required
              error={!formData.state}
              label="State"
              defaultValue={address.state}
              onChange={(event) => {
                setFormData({ ...formData, state: event.target.value });
              }}
            />
            <TextField
              required
              error={!formData.zip}
              label="ZIP"
              type="number"
              defaultValue={address.zip}
              onChange={(event) => {
                setFormData({ ...formData, zip: parseInt(event.target.value) });
              }}
            />
          </div>
        </Box>
      ) : (
        <Stack>
          <Typography variant="body2" component="div">
            {address.line1}
          </Typography>
          {address.line2 !== "" && (
            <Typography variant="body2" component="div">
              {address.line2}
            </Typography>
          )}
          <Typography variant="body2" component="div">
            {`${address.city}, ${address.state} ${address.zip}`}
          </Typography>
        </Stack>
      )}
      <Button
        style={{ width: 15, marginTop: 20 }}
        variant="contained"
        onClick={handleClick}
        type="submit"
      >
        {isEditing ? "Save" : "Edit"}
      </Button>
    </Card>
  );
}
