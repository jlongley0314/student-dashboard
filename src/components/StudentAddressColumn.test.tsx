import React from "react";
import {
  render,
  RenderResult,
  screen,
  fireEvent,
} from "@testing-library/react";
import { StudentAddressColumn } from "./StudentAddressColumn";
import { useEditAddress } from "../mutations/useEditAddress";
import { QueryClient, QueryClientProvider } from "react-query";
jest.mock("../mutations/useEditAddress", () => ({
  useEditAddress: jest.fn(),
}));

describe("Header", () => {
  let component: RenderResult;
  const id = "1";
  const address = {
    id: "111",
    line1: "1234 Tracy Ave",
    line2: "3A",
    city: "Kansas City",
    state: "MO",
    zip: 64110,
  };
  let mutateMock: jest.Mock;

  beforeEach(() => {
    mutateMock = jest.fn();
    (useEditAddress as jest.Mock).mockReturnValue({
      mutate: mutateMock,
      isSuccess: true,
    });
    const queryClient = new QueryClient();
    component = render(
      <QueryClientProvider client={queryClient}>
        <StudentAddressColumn
          id={id}
          address={address}
          addressLoading={false}
        />
      </QueryClientProvider>
    );
  });

  it("should render component as expected", () => {
    expect(component).toMatchSnapshot();
  });

  it("should show the edit form when edit button is selected", () => {
    expect(screen.queryByTestId("form-box")).toBeFalsy();
    fireEvent.click(screen.getByText("Edit"));
    expect(screen.getByTestId("form-box")).toBeDefined();
  });

  it("should call edit address mutation with correct form data when submit button is pressed", () => {
    fireEvent.click(screen.getByText("Edit"));
    fireEvent.click(screen.getByText("Save"));
    expect(mutateMock).toBeCalledWith(address);
  });
});
