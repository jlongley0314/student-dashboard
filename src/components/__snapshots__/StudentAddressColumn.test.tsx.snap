// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render component as expected 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
        style="padding: 10px;"
      >
        <div
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        >
          Address:
        </div>
        <div
          class="css-nen11g-MuiStack-root"
        >
          <div
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            1234 Tracy Ave
          </div>
          <div
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            3A
          </div>
          <div
            class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          >
            Kansas City, MO 64110
          </div>
        </div>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          style="width: 15px; margin-top: 20px;"
          tabindex="0"
          type="submit"
        >
          Edit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      style="padding: 10px;"
    >
      <div
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
      >
        Address:
      </div>
      <div
        class="css-nen11g-MuiStack-root"
      >
        <div
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          1234 Tracy Ave
        </div>
        <div
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          3A
        </div>
        <div
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          Kansas City, MO 64110
        </div>
      </div>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        style="width: 15px; margin-top: 20px;"
        tabindex="0"
        type="submit"
      >
        Edit
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
